[Concepts]: Strings (#378)
exercism/crystal: concepts-wip/strings/introduction.md

# Strings

Crystal has a type known as [`String`][strings] which is used to represent text.

> In Crystal is a String a sequence of unicode characters.

Reviewer: ```suggestion
Crystal strings are sequences of unicode characters.
```

Strings are immutable, meaning they cannot be changed once created.
This means that every time you do a modification on a string a new string is created, thereby the original string is not modified.

To define a string you can use double quotes `"`.

```crystal
"Hello World"
```

## Concatenation

To concatenate two strings you can use the `+` operator.
The `+` operator will return a new string that is the combination of the two strings given.

```crystal
hello = "Hello"
hello + " World"
# => "Hello World"
```

> 
> When wanting to concat a string with a non-string type so can you use the method `to_s` to convert the type to a string.
> 
> ```crystal
> "Hello " + 1.to_s
> # => "Hello 1"
> ```

Reviewer: ```suggestion
```
You talk about `to_s` just below.

If the string can't be converted to a number then an ArgumentError will be raised.

```crystal
"1".to_i
# => 1

"1.0".to_f
# => 1.0

"Hello".to_i
# => ArgumentError: Invalid integer: Hello
```

An integer or floating point number can be converted to a string using the `to_s` method.

```crystal
1.to_s
# => "1"

1.0.to_s
# => "1.0"
```

## Interpolation

Interpolation is a convenient way to combine strings and embeds expressions in strings.
To interpolate a string you can use the `#` character followed by curly braces `{}` with the expression inside the braces.

```crystal
name = "World"
"Hello #{name}!"
# => "Hello World!"
```

Crystal will automatically convert the result of the expression to a string.

```crystal
"Hello #{1 + 1}!"
# => "Hello 2!"
```

## Size

When you need to know the number of characters in a string you can use the `size` method, which will return the length of the string as an integer.
The size of a string is a stored property of the string, so it doesn't have to calculate the size every time you call the method, making it very fast.

```crystal
"Hello World".size
# => 11
```

## Indexing

Indexing is when you want to get a specific character from a string.

> To get a character from a string you can use the `[]` method with the index of the character you want to get.

Reviewer: A suggestion:

```suggestion
To get a character from a string you can use familiar bracket notation.

~~~~exercism/note
`[]` is actually implemented as a String instance method, where the index is the method argument.
~~~~
```
